sklearn_numba_dpex/examples/kmeans_numba_dpex.py

```diff
@@ -0,0 +1,628 @@
+                          n_clusters=128,
+                          init="random",
+                          n_init=10,
+                          max_iter=300,
+                          tol=1e-4,
+                          random_state=123,
+                          algorithm="full",
+                          copy_x=False):
+
+    #!!!: x_train, y_train must be in row_major order (C-style)
+    # and should be float32
+
+    # TODO: write a kernel for this ?
+    # x_train[:] = x_train - x_train.mean(axis=0)
+
+    # TODO: use SYCL array rather than numpy arrays
+    # enables finer control of data location
+    dim = x_train.shape[1]
+    n = x_train.shape[0]
+
+    # TODO: write a kernel for initialization on GPU
+    # implement random AND ++
+    rng = default_rng(random_state)
+    centroids = np.array(rng.choice(x_train, n_clusters, replace=False), dtype=np.float32).copy()
+
+    # TODO: the conditions used here to select the best kernel are the conditions
+    # suggested in the paper, but it seems that the conditions are slightly
+    # different when reading the implementation of the authors. Find the
+    # differences between the two and use the appropriate best conditions.
+    if n_clusters <= 32 or dim >= 64:
+        h = 16
+        r = 32
+        thread_group_size = 256
+        n_cluster_groups, assignment_step = get_assignment_step_fixed_kernel(
+            h, r, thread_group_size, n, dim, n_clusters)
+    else:
+        rq = 4
+        rp = 4
+        p = 32
+        q = 4
+        n_cluster_groups, assignment_step = get_assignment_step_regs_kernel(
+            rp, rq, p, q, n, dim, n_clusters)
+
+    coarse_assignments_dist = np.empty((n, n_cluster_groups), dtype=np.float32)
+    coarse_assignments_idx = np.empty((n, n_cluster_groups), dtype=np.uint32)
+
+    assignment_step(x_train, centroids, coarse_assignments_dist, coarse_assignments_idx)
+
+    # fine_assignments_thread_group_size = 256
+    # fine_assignments_n_threads = np.ceil(n/256) * 256
+
+    # if n_cluster_groups > 1:
+    #     fine_assignments_dist = np.empty(n, dtype=np.float32)
+    #     fine_assignments_idx = np.empty(n, dtype=np.uint32)
+
+    # while True:
+    #     import ipdb; ipdb.set_trace()
+    #     assignment_step(x_train, centroids, coarse_assignments_dist, coarse_assignments_idx)
+    #     if n_cluster_groups > 1:  # still TODO
+    #         fine_assignment_step[
+    #             fine_assignments_n_threads, fine_assignments_thread_group_size](
+    #                 coarse_assignments_dist, coarse_assignments_idx,
+    #                 fine_assignments_dist, fine_assignments_idx)
+    #     else:
+    #         fine_assignments_dist = coarse_assignments_dist[:]
+    #         fine_assignments_idx = coarse_assignments_idx[:]
+
+    # update_means = (update_means_step_fine if dim >= 8192 else update_means_step_shm)
+
+
+if __name__ == "__main__":
+    x, y = fetch_openml(name='spoken-arabic-digit', return_X_y=True)
```

Comment:
Upper-casing the first letter is used conventionally for scikit-learn for dataset.
```suggestion
    X, y = fetch_openml(name='spoken-arabic-digit', return_X_y=True)
```

This applies here and in all other places.